### PR TITLE
Annotation list: sort by date

### PR DESCRIPTION
**Pull request by hius07**
- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12020)

### CODE REVIEW

File: frontend/apps/reader/modules/readerbookmark.lua [hunk 1]
```diff
@@ -264,6 +275,35 @@ function ReaderBookmark:genShowInItemsMenuItems(value)
...
 end
 
+function ReaderBookmark:genSortByMenuItems(value, separator)
+    local strings = {
+        page = _("page number"),
+        date = _("date"),
+    }
+    local strings_reverse = {
+        page = _("page number, reverse"),
+        date = _("date, reverse"),
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerbookmark.lua:285]: Ascending/descending are often used, but I find reverse clearer. ;-)

**Review by Frenzie — APPROVED** [review]

**Comment by hius07** [comment]
Excuse me, I am deliberately breaking the backward compatibility: new users and users without `bookmarks_items_reverse_sorting` in the settings will get non-reverse order.

**Review by poire-z — APPROVED** [review]
Trusting you on the code.

**Comment by poire-z** [comment]
> Excuse me, I am deliberately breaking the backward compatibility: new users and users without bookmarks_items_reverse_sorting in the settings will get non-reverse order.

Why ? For technical reasons, or because you think it is better non-reverse ?
(Not often using bookmarks, but it feels better to have them in reverse with the newest at top - rather than in order and jumping on the last page, and finding them either at top, in the middle, at bottom. No strong opinion though.)

**Comment by hius07** [comment]
Not technical reasons.
(1) If bm are sorted by page number, the bm list is opened on the page containing the bookmark in the current book page (or the closest one). So if we read forward and make bookmarks - we always see the latest bookmark on opening the bm list. (Just to remind that page numbers of bookmarks after the current page are dimmed). And it is convenient to see the page numbers ascending like in the ToC.
(2) If bm are sorted by date (new mode), the non-reverse ordering means the date descending, i.e. the first bookmark in the list is the most recent one. And in this mode bm list is opened in the first page:
https://github.com/koreader/koreader/blob/b15cf2967112371ea897b4e64e8e94e317d6f8f6/frontend/apps/reader/modules/readerbookmark.lua#L974

Summarizing, I believe that non-reverse mode should be the default one, and the setting for it is saved as nil.

**Comment by poire-z** [comment]
Fine with your explanations.

> (2) If bm are sorted by date (new mode), the **non-reverse** ordering means the date descending, i.e. the **first bookmark in the list is the most recent one**. And in this mode bm list is opened in the first page:

This feels a bit odd. that the `[ ] reverse` would have quite opposite effects whether by page / by date. For someone that reads naturally from book start to end, I would expect these 2 options would behave quite the same.

**Comment by hius07** [comment]
File browser sort by date shows the most recent first.

**Comment by hius07** [comment]
@ryanwwest, what is the non-reverse order for sorting bookmarks by date: ascending or descending?

**Comment by ryanwwest** [comment]
@hius07 I think descending would be first page of results top result has the highest number (most recent), then gets further away. Ascending would then be oldest entries first. Though not sure if I understood.

I notice that some bookmarks options are found in the Bookmarks list page in the hamburger menu item (filtering) while others are in Bookmark icon -> Settings -> Bookmarks. I was thinking that sort would make more sense to go with the Filter content since the typical online workflow with searching things is to have filter and sort options nested together. But I also see that 'Sort by largest page number' is already not with the filter items. I wonder if it would be better to move the sort options to the filter menu?

And at a bigger level, potentially unify all the Bookmarks page list settings in one place but that's probably for another time and debate.

**Comment by hius07** [comment]
So, "Sort by date" shows the oldest bookmark first in the list, and "Sort by date, reverse" shows the most recent bookmark first, right?
Do you want to see the bookmark date in the list? Where? I'd like to show the page number in any sorting mode. Adding the datetime to the page number would eat the room for bookmark text.

**Comment by hius07** [comment]
I'm also going to add searching by datetime in `2024-04-06 22:48:28` format (can be partial). Is it useful?

**Comment by ryanwwest** [comment]
Thanks for this, @hius07 .

If you're asking my opinion (I haven't tested the PR), while I think most recent bookmarks first would be helpful, I get @poire-z 's point that maybe bookmark order should be default (which shouldn't say 'reversed') match earliest to latest like when you read a book left to right, earliest to latest. I think people can just change the order as long as it's easy for them to find sort setting.

As for date visibility in Bookmarks list, yes that's for #2 of https://github.com/koreader/koreader/issues/11822. Could it go under the page number on the right, so it's always visible and in the same place even with long note/highlight? I'm not sure how much space that needs.

I think that's the best time format. When you say search, is this search restricted to bookmarks page only or more general?

**Comment by ryanwwest** [comment]
Maybe for timestamp location in bookmark list, if it's right-aligned below the page number the date could be on top line, time on bottom line, so it take up less horizontal space.

**Comment by hius07** [comment]
Our right part (tab) of the list is single-line only, sorry.

**Comment by jonnyl2** [comment]
>>[hius07](2) If bm are sorted by date (new mode), the non-reverse ordering means the date descending, i.e. the first bookmark in the list is the most recent one. And in this mode bm list is opened in the first page:

>[poire-z] This feels a bit odd. that the [ ] reverse would have quite opposite effects whether by page / by date. For someone that reads naturally from book start to end, I would expect these 2 options would behave quite the same.

I think there was a misunderstanding here? They do behave the same as @hius07 described: non-reverse order means the latest bookmark OR the highest page number comes first. And in my opinion that is the most desirable behavior as well. 

To summarize: Non-reverse by date/time: 22:00; 21:33; 20:25...  |  by page: 800; 677; 533... 
Reverse: 20:25; 21:33; 22:00 | 533; 677; 800

**Comment by hius07** [comment]
Changes:
(1) consistent "page" and "date" sorting modes: non-reverse mode means page increasing or date from old to new.
(2) in "date" sorting mode append the datetime in the beginning of the item text text

**Pull request merged**